test/e2e_node/deleted_pods_test.go

```diff
@@ -215,4 +216,106 @@ var _ = SIGDescribe("Deleted pods handling", framework.WithNodeConformance(), fu
 	)
 
+	ginkgo.It("Should report true container state when pod have terminationGracePeriodSeconds", func(ctx context.Context) {
+		podName := "deletion-grace-period-pod" + string(uuid.NewUUID())
+		podSpec := e2epod.MustMixinRestrictedPodSecurity(&v1.Pod{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:       podName,
+				Finalizers: []string{testFinalizer},
+			},
+			Spec: v1.PodSpec{
+				RestartPolicy:                 v1.RestartPolicyNever,
+				TerminationGracePeriodSeconds: &[]int64{60}[0],
+				Containers: []v1.Container{
+					{
+						Name:    "c1",
+						Image:   imageutils.GetE2EImage(imageutils.BusyBox),
+						Command: []string{"/bin/sh", "-c"},
+						Args: []string{
+							`_term() {
+            rm -f /tmp/ready
+          }
+          trap _term SIGTERM
+          
+          touch /tmp/ready
+          
+          while true; do
+            echo 'helloc1'
+            ls /tmp/die_now && echo 'dying in 5s...' && sleep 5 && exit 0
+            sleep 1
+          done`,
+						},
+						ReadinessProbe: &v1.Probe{
+							ProbeHandler: v1.ProbeHandler{
+								Exec: &v1.ExecAction{
+									Command: []string{"/bin/sh", "-c", `if [ -f "/tmp/ready" ]; then
+                exit 0
+              else
+                touch /tmp/die_now
+                exit 1
+              fi`},
+								},
+							},
+						},
+					},
+					{
+						Name:    "c2",
+						Image:   imageutils.GetE2EImage(imageutils.BusyBox),
+						Command: []string{"/bin/sh", "-c", `_term() { while true; do echo \"hello_term_c2\"; sleep 1; done } ; trap _term SIGTERM; while true; do echo \"helloc2\"; sleep 1; done`},
+					},
+				},
+			},
+		})
+		ginkgo.By(fmt.Sprintf("Creating pod %s/%s", f.Namespace.Name, podSpec.Name))
+		pod := e2epod.NewPodClient(f).Create(ctx, podSpec)
+
+		ginkgo.By("set up cleanup of the finalizer")
+		ginkgo.DeferCleanup(e2epod.NewPodClient(f).RemoveFinalizer, pod.Name, testFinalizer)
+
+		ginkgo.By("waiting for pod to be running")
+		err := e2epod.WaitTimeoutForPodReadyInNamespace(ctx, f.ClientSet, pod.Name, f.Namespace.Name, f.Timeouts.PodStart)
+		framework.ExpectNoError(err, "failed to wait for pod to be ready")
+
+		ginkgo.By(fmt.Sprintf("Deleting the pod (%v/%v) with terminationGracePeriodSeconds to set a deletion timestamp", f.Namespace.Name, pod.Name))
+		err = e2epod.NewPodClient(f).Delete(ctx, pod.Name, metav1.DeleteOptions{})
+		framework.ExpectNoError(err, "failed to delete the pod: %q", pod.Name)
+
+		// Since TerminationGracePeriodSeconds is set to 60, c1 will be deleted, while c2 will not be deleted immediately.
+		// Therefore, before c2 is deleted, the ready status of c1 is false, and the ready status of c2 is true.
+		// detail see #129552
+		gomega.Eventually(ctx, func() bool {
+			pod, err := e2epod.NewPodClient(f).Get(ctx, pod.Name, metav1.GetOptions{})
+			framework.ExpectNoError(err, "failed to get the pod: %q", pod.Name)
+
+			var c1Ready, c2Ready bool
```

Comment:
nit: I think that instead of using for loop, it might be fine to check ready state by directly specifying the index of container statuses, as done in lines 314–318.

Reply:
done